ghcide/src/Development/IDE/Types/Logger.hs

```diff
@@ -57,3 +91,180 @@ logTelemetry x = logPriority x Telemetry
 
 noLogging :: Logger
 noLogging = Logger $ \_ _ -> return ()
+
+data WithPriority a = WithPriority { priority :: Priority, payload :: a } deriving Functor
+
+-- | Note that this is logging actions _of the program_, not of the user.
+--   You shouldn't call warning/error if the user has caused an error, only
+--   if our code has gone wrong and is itself erroneous (e.g. we threw an exception).
+data Recorder msg = Recorder
+  { logger_ :: forall m. (HasCallStack, MonadIO m) => msg -> m () }
```

Comment:
`Recorder` is a `Semigroup` where `<>` is used to compose multiple backends. It's also a `Monoid`.

Reply:
I see, so another argument to have a mega `Log` type is so we can compose backends easily.

Reply:
Not at all. What makes you think that?

Top level logger from Sigma IDE in case it helps:
```
  logger =
    VLog.logger (priorityFunction opts)
      <> scubaTracer
      <> (toTelemetry >$< eventlogTracer)
      <> if testInterfaceLoading opts
        then enforceInterfaceLoadingLogger myTid
        else mempty
```

Reply:
I think what made me think that is if we have
```
(<>) :: Logger msg -> Logger msg -> Logger msg
```
then `msg` need to be the same for both arguments.

If we have a logger that sends text to console, and a logger that sends json to client then maybe we start with
`Logger Text` and `Logger Json`, and so in order to make `Text` and `Json` the same we need to contramap them to the same type. To do this for all possible log values in the program we can embed the log values in each other like an exception hierarchy so that at some point every possible log value is contained somewhere under a `Log` type. Then we can `(<>)` the loggers that have been contramapped to this `Log` type.

Without a final `Log` type we could still combine backends but more locally, and not in a single place.

Reply:
I think you might be talking about different things. I think Pepe maybe means "backends" to just be "recorders", in which case it already *has* those `Semigroup`/`Monoid` instances, I think?

Reply:
Oh, right, the instances are already there below a long comment. Sorry, should have scrolled down further

Reply:
@eddiemundo exactly, I guess this doesn't mean a mega `Log` type to me. It's a composable `Log` type.